genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.rpc.grpc.services.impl.v4;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.netflix.genie.common.internal.dto.v4.files.JobFileState;
+import com.netflix.genie.proto.BeginAcknowledgement;
+import com.netflix.genie.proto.BeginSync;
+import com.netflix.genie.proto.DataUpload;
+import com.netflix.genie.proto.JobDirectoryState;
+import com.netflix.genie.proto.JobFileSyncServiceGrpc;
+import com.netflix.genie.proto.ResetSync;
+import com.netflix.genie.proto.SyncComplete;
+import com.netflix.genie.proto.SyncRequest;
+import com.netflix.genie.proto.SyncResponse;
+import com.netflix.genie.proto.UploadAcknowledgement;
+import com.netflix.genie.web.services.JobFileService;
+import io.grpc.stub.StreamObserver;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+import org.apache.commons.lang.StringUtils;
+import org.springframework.scheduling.TaskScheduler;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of the gRPC Job File Sync interface for syncing job files from agent to server via bi-directional
+ * connection.
+ *
+ * @author tgianos
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobFileSyncServiceGrpc.class
+//    interceptors = {
+//        SimpleLoggingInterceptor.class,
+//    }
+)
+@Slf4j
+public class GRpcJobFileSyncServiceImpl extends JobFileSyncServiceGrpc.JobFileSyncServiceImplBase {
+
+    private final JobFileService jobFileService;
+    private final TaskScheduler taskScheduler;
+
+    private final Map<String, GenieJobFileSyncObserver> jobSyncRequestObservers = Maps.newConcurrentMap();
```

Comment:
Instead of having  `Maps.newConcurrentMap();`, would it be better to inject this as a dependency. It might help in testing. I had a similar map of observers for `JobKillService` and I was thinking of creating a interface and injecting its Impl

Reply:
I don't agree with this. This is an internal construct for keeping state important to this particular implementation. If you provide a map externally you're effectively risking manipulation of that Map by outside sources, not to mention exposing internal implementation and breaking encapsulation. No one else should really care that you're keeping this map it's just for internal housekeeping for this class. If you really need access for it for testing you can provide a `protected` or `visablefortesting` getter to check the contents of the map.